File: app/src/main/java/com/lz233/onetext/service/ExternalControlService.java
package com.lz233.onetext.service;

import android.app.Activity;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.service.controls.Control;
import android.service.controls.ControlsProviderService;
import android.service.controls.DeviceTypes;
import android.service.controls.actions.ControlAction;
import android.service.controls.templates.ControlButton;
import android.service.controls.templates.ControlTemplate;
import android.service.controls.templates.RangeTemplate;
import android.service.controls.templates.StatelessTemplate;
import android.service.controls.templates.TemperatureControlTemplate;
import android.service.controls.templates.ToggleRangeTemplate;
import android.service.controls.templates.ToggleTemplate;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import com.lz233.onetext.activity.MainActivity;

import org.reactivestreams.FlowAdapters;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Flow;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.processors.ReplayProcessor;

@RequiresApi(api = Build.VERSION_CODES.R)
public class ExternalControlService extends ControlsProviderService {
    private ReplayProcessor updatePublisher;
    private final List<Control> mAllControls = new ArrayList<>();
    private PendingIntent mPendingIntent = PendingIntent.getActivity(getBaseContext(), 1, new Intent(), PendingIntent.FLAG_UPDATE_CURRENT);
    @NonNull
    @Override
    public Flow.Publisher<Control> createPublisherForAllAvailable() {
        return new CtsControlsPublisher(mAllControls.stream()
                .map(c -> new Control.StatelessBuilder(c).build())
                .collect(Collectors.toList()));
    }

    @NonNull
    @Override
    public Flow.Publisher<Control> createPublisherFor(@NonNull List<String> list) {
        Context context = getBaseContext();
        /* Fill in details for the activity related to this device. On long press,
         * this Intent will be launched in a bottomsheet. Please design the activity
         * accordingly to fit a more limited space (about 2/3 screen height).
         */
        Intent i = new Intent();
        PendingIntent pi = PendingIntent.getActivity(context, 0, i, PendingIntent.FLAG_UPDATE_CURRENT);

        updatePublisher = ReplayProcessor.create();

        // For each controlId in controlIds

        if (list.contains("233")) {

            Control control = new Control.StatefulBuilder("233", pi)
                    // Required: The name of the control
                    .setTitle("MY-CONTROL-TITLE")
                    // Required: Usually the room where the control is located
                    .setSubtitle("MY-CONTROL-SUBTITLE")
                    // Optional: Structure where the control is located, an example would be a house
                    .setStructure("MY-CONTROL-STRUCTURE")
                    // Required: Type of device, i.e., thermostat, light, switch
                    .setDeviceType(DeviceTypes.TYPE_UNKNOWN) // For example, DeviceTypes.TYPE_THERMOSTAT
                    // Required: Current status of the device
                    .setStatus(Control.STATUS_UNKNOWN)
                    .setControlTemplate(new ToggleTemplate("2333", new ControlButton(false,"action")))// For example, Control.STATUS_OK
                    .build();

            updatePublisher.onNext(control);
        }
        // Uses the Reactive Streams API
        return FlowAdapters.toFlowPublisher(updatePublisher);
    }

    @Override
    public void performControlAction(@NonNull String s, @NonNull ControlAction controlAction, @NonNull Consumer<Integer> consumer) {

    }
    public Control buildLight(boolean isOn, float intensity) {
        RangeTemplate rt = new RangeTemplate("range", 0.0f, 100.0f, intensity, 1.0f, null);
        ControlTemplate template =
                new ToggleRangeTemplate("toggleRange", isOn, isOn ? "On" : "Off", rt);
        return new Control.StatefulBuilder("light", mPendingIntent)
                .setTitle("Light Title")
                .setSubtitle("Light Subtitle")
                .setStatus(Control.STATUS_OK)
                .setStatusText(isOn ? "On" : "Off")
                .setDeviceType(DeviceTypes.TYPE_LIGHT)
                .setStructure("Home")
                .setControlTemplate(template)
                .build();
    }

    public Control buildSwitch(boolean isOn) {
        ControlButton button = new ControlButton(isOn, isOn ? "On" : "Off");
        ControlTemplate template = new ToggleTemplate("toggle", button);
        return new Control.StatefulBuilder("switch", mPendingIntent)
                .setTitle("Switch Title")
                .setSubtitle("Switch Subtitle")
                .setStatus(Control.STATUS_OK)
                .setStatusText(isOn ? "On" : "Off")
                .setDeviceType(DeviceTypes.TYPE_SWITCH)
                .setStructure("Home")
                .setControlTemplate(template)
                .build();
    }
}

File: app/src/main/java/com/lz233/onetext/service/ExternalControlService.kt
package com.lz233.onetext.service

import android.app.PendingIntent
import android.content.Intent
import android.graphics.drawable.Icon
import android.os.Build
import android.service.controls.Control
import android.service.controls.ControlsProviderService
import android.service.controls.DeviceTypes
import android.service.controls.actions.ControlAction
import android.service.controls.templates.StatelessTemplate
import androidx.annotation.RequiresApi
import com.lz233.onetext.OneTextApp
import com.lz233.onetext.R
import com.lz233.onetext.activity.MainActivity
import com.lz233.onetext.tools.utils.CoreUtil
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.processors.ReplayProcessor
import org.reactivestreams.FlowAdapters
import java.util.*
import java.util.concurrent.Flow
import java.util.function.Consumer
import kotlin.collections.HashMap

@RequiresApi(Build.VERSION_CODES.R)
class ExternalControlService : ControlsProviderService() {
    val context = OneTextApp.context
    val i = Intent(context, MainActivity::class.java).apply {
        flags = Intent.FLAG_ACTIVITY_NEW_TASK
        putExtra("fromExternalControl", true)
    }
    val pi: PendingIntent =
            PendingIntent.getActivity(
                    context, 2333, i,
                    PendingIntent.FLAG_UPDATE_CURRENT
            )
    val updatePublisher = ReplayProcessor.create<Control>()
    override fun createPublisherForAllAvailable(): Flow.Publisher<Control> {
        val controls = mutableListOf(buildOnetext())
        // Create more controls here if needed and add it to the ArrayList

        // Uses the RxJava 2 library
        return FlowAdapters.toFlowPublisher(Flowable.fromIterable(controls))
    }

    override fun createPublisherFor(controlIds: MutableList<String>): Flow.Publisher<Control> {
        if (controlIds.contains("onetext")) {
            val control = buildOnetext()

            updatePublisher.onNext(control)
        }

        // If you have other controls, check that they have been selected here

        // Uses the Reactive Streams API
        return FlowAdapters.toFlowPublisher(updatePublisher)
    }

    override fun performControlAction(controlId: String, action: ControlAction, consumer: Consumer<Int>) {
        /* First, locate the control identified by the controlId. Once it is located, you can
         * interpret the action appropriately for that specific device. For instance, the following
         * assumes that the controlId is associated with a light, and the light can be turned on
         * or off.
         */
        if (controlId == "onetext") {

            // Inform SystemUI that the action has been received and is being processed
            consumer.accept(ControlAction.RESPONSE_OK)
            // In this example, action.getNewState() will have the requested action: true for “On”,
            // false for “Off”.

            /* This is where application logic/network requests would be invoked to update the state of
             * the device.
             * After updating, the application should use the publisher to update SystemUI with the new
             * state.
             */
            // This is the publisher the application created during the call to createPublisherFor()
            updatePublisher.onNext(buildOnetext(true))
        }
    }

    private fun buildOnetext(forceFresh: Boolean = false): Control {
        val coreUtil = CoreUtil(context)
        var hashMap: HashMap<Any, Any> = try {
            coreUtil.getOneText(forceFresh, true)
        }catch (e: Throwable){
            e.printStackTrace()
            HashMap<Any, Any>().apply {
                put("text", getString(R.string.control_long_press))
                put("by", "")
            }
        }
        //更新小部件
        context.sendBroadcast(Intent("com.lz233.onetext.widget").apply { setPackage(packageName) })
        return Control.StatefulBuilder("onetext", pi)
                .setTitle(hashMap["text"] as String)
                .setSubtitle(hashMap["by"] as String)
                .setStructure(getString(R.string.app_name))
                .setDeviceType(DeviceTypes.TYPE_UNKNOWN)
                .setStatus(Control.STATUS_OK)
                .setStatusText(getString(R.string.app_name))
                .setControlTemplate(StatelessTemplate("onetext"))
                .setCustomIcon(Icon.createWithResource(OneTextApp.context, R.drawable.ic_notification).setTint(getColor(R.color.colorWhiteWidget)))// For example, DeviceTypes.TYPE_THERMOSTAT
                .build()
    }
}